core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -121,32 +227,36 @@ impl ClusterSlots {
             slot_range.start > current_slot,
             "Can not roll cluster slots backwards!"
         );
-        let map_capacity = validator_stakes.len();
         let mut cluster_slots = self.cluster_slots.write().unwrap();
         self.metric_write_locks.fetch_add(1, Ordering::Relaxed);
-        // zero-initialize a given map
-        let zero_init_map = |map: &mut SlotPubkeys| {
-            map.extend(
-                validator_stakes
-                    .iter()
-                    .map(|(k, _v)| (*k, AtomicU64::new(0))),
-            );
-        };
-        // a handy closure to spawn new hashmaps with the correct parameters
-        let map_maker = || {
-            let mut map =
-                HashMap::with_capacity_and_hasher(map_capacity, PubkeyHasherBuilder::default());
-            zero_init_map(&mut map);
-            self.metric_allocations.fetch_add(1, Ordering::Relaxed);
-            map
-        };
+        let epoch_metadata = self.epoch_metadata.read().unwrap();
         //startup init, this is very slow but only ever happens once
         if cluster_slots.is_empty() {
+            info!("Init cluster_slots at range {:?}", slot_range);
             for slot in slot_range.clone() {
+                // Epoch should be defined for all slots in the window
+                let epoch = self
+                    .get_epoch_for_slot(slot)
+                    .expect("Epoch should be defined for all slots");
+                let supporters = if let Some(epoch_data) = epoch_metadata.get(&epoch) {
```

Comment:
Epoch data might actually not be available if epochs are too short. In this case we fill ringbuffer with blanks.